### PR TITLE
Add a simple helm test

### PR DESCRIPTION
**Pull request by aLekSer**
This could be useful for Terraform configs check that installation
was completed, for example after switching to a new cluster.

Run `helm test agones --cleanup` to verify the deployment.

### CODE REVIEW

**Comment by aLekSer** [comment]
In Helm 3 there is an ability to use Batch Job instead of Pod in `test-runner.yaml`
We can add {{- if .Values.tests.enabled }} parameter as in this example https://github.com/hashicorp/consul-helm/blob/master/templates/tests/test-runner.yaml

**Comment by aLekSer** [comment]
Build failed because of absence of the image.

File: examples/crd-client/main.go [hunk 1]
```diff
@@ -15,8 +15,10 @@
 package main
 
 import (
-	"fmt"
 	"strings"
+	"time"
+
```
Comment (markmandel) [inline examples/crd-client/main.go:20]: Nit: cleanup spacing.

**Comment by aLekSer** [comment]
Yes, I will add this to the documentation.

**Comment by markmandel** [comment]
Gentle bump to add documentation.

File: examples/crd-client/main.go [hunk 2]
```diff
@@ -79,15 +100,27 @@ func main() {
...
+	newGS, err := agonesClient.AgonesV1().GameServers(defaultNs).Create(gs)
 	if err != nil {
 		panic(err)
 	}
+	logrus.Infof("New GameServer name is: %s", newGS.ObjectMeta.Name)
 
-	fmt.Printf("New game servers' name is: %s", newGS.ObjectMeta.Name)
+	if viper.GetBool(isHelmTest) {
+		time.Sleep(1 * time.Second)
+		err = agonesClient.AgonesV1().GameServers(defaultNs).Delete(newGS.ObjectMeta.Name, nil) // nolint: errcheck
```
Comment (akremsa) [inline examples/crd-client/main.go:121]: remove // nolint: errcheck

File: examples/crd-client/main.go [hunk 3]
```diff
@@ -79,15 +100,27 @@ func main() {
...
+		if err != nil {
+			panic(err)
```
Comment (akremsa) [inline examples/crd-client/main.go:123]: It is better to use panic from logrus here and above

File: examples/crd-client/main.go [hunk 4]
```diff
@@ -30,19 +31,30 @@ import (
...
 	viper.AllowEmptyEnv(true)
 	viper.SetEnvKeyReplacer(strings.NewReplacer("-", "_"))
 
-	pflag.String(gameServerImage, viper.GetString(gameServerImage), "The Address to bind the server grpcPort to. Defaults to 'localhost'")
+	pflag.String(gameServerImage, defaultImage, "The Address to bind the server grpcPort to. Defaults to 'localhost'")
 	viper.SetDefault(gameServerImage, defaultImage)
 	runtime.Must(viper.BindEnv(gameServerImage))
-	viper.GetString(gameServerImage)
+
+	pflag.Bool(isHelmTest, false, "Is helm test - shutdown GameServer at the end of test. Defaults to false")
```
Comment (akremsa) [inline examples/crd-client/main.go:50]: Can we rephrase - "Defines whether GameServer should be shut down at the end of the test or not"?

Comment (aLekSer) [reply examples/crd-client/main.go:50]: added your definition.

File: examples/crd-client/main.go [hunk 5]
```diff
@@ -66,8 +83,14 @@ func main() {
 		logger.WithError(err).Fatal("Could not create the agones api clientset")
 	}
 
+	gsName := "helm-test-server-"
```
Comment (akremsa) [inline examples/crd-client/main.go:86]: Minor: I see that this variable is used only in 91 line, so let's make this code a bit cleaner and move this value either to gs creation step or make it constant and move to the top of the file.

**Comment by aLekSer** [comment]
E2E test fail:
```
time="2020-04-21 11:33:07.597" level=info msg="Finished cleanup."
--- FAIL: TestAllocator (1.45s)
    allocator_test.go:61: deleting pods failed: pods "agones-allocator-65f5dfb9db-s9n7s" not found
time="2020-04-21 11:33:12.564" level=info msg="waiting for fleet condition" fleet=simple-fleet-66c4t
time="2020-04-21 11:33:12.652" level=info msg="fleet simple-fleet-66c4t has 0/3 ready replicas"
time="2020-04-21 11:33:14.713" level=info msg="fleet simple-fleet-66c4t has 0/3 ready replicas"
time="2020-04-21 11:33:16.717" level=info msg="fleet simple-fleet-66c4t has 0/3 ready replicas"
time="2020-04-21 11:33:18.714" level=info msg="fleet simple-fleet-66c4t has 3/3 ready replicas"
time="2020-04-21 11:33:18.923" level=info msg="failing PostAllocate request" error="rpc error: code = Unavailable desc = all SubConns are in TransientFailure, latest connection error: connection error: desc = \"transport: authentication handshake failed: x509: cannot validate certificate for 35.247.4.172 because it doesn't contain any IP SANs\""
```

**Comment by aLekSer** [comment]
Added docs about how to run `helm test agones --cleanup`. Now it also wait for `GameServer` to become `Ready`.

**Comment by aLekSer** [comment]
Fail in C# SDK conformance test:
```

/usr/share/dotnet/sdk/2.2.402/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.targets(129,5): error MSB4018: The "GenerateDepsFile" task failed unexpectedly. [/go/src/agones.dev/agones/sdks/csharp/sdk/csharp-sdk.csproj]
/usr/share/dotnet/sdk/2.2.402/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.targets(129,5): error MSB4018: System.IO.IOException: The process cannot access the file '/go/src/agones.dev/agones/sdks/csharp/sdk/bin/Release/netstandard2.0/AgonesSDK.deps.json' because it is being used by another process. [/go/src/agones.dev/agones/sdks/csharp/sdk/csharp-sdk.csproj]
/usr/share/dotnet/sdk/2.2.402/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.targets(129,5): error MSB4018:    at System.IO.FileStream.Init(FileMode mode, FileShare share) [/go/src/agones.dev/agones/sdks/csharp/sdk/csharp-sdk.csproj]
/usr/share/dotnet/sdk/2.2.402/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.targets(129,5): error MSB4018:    at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize, FileOptions options) [/go/src/agones.dev/agones/sdks/csharp/sdk/csharp-sdk.csproj]
/usr/share/dotnet/sdk/2.2.402/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.targets(129,5): error MSB4018:    at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize) [/go/src/agones.dev/agones/sdks/csharp/sdk/csharp-sdk.csproj]
/usr/share/dotnet/sdk/2.2.402/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.targets(129,5): error MSB4018:    at System.IO.File.Create(String path) [/go/src/agones.dev/agones/sdks/csharp/sdk/csharp-sdk.csproj]
/usr/share/dotnet/sdk/2.2.402/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.targets(129,5): error MSB4018:    at Microsoft.NET.Build.Tasks.GenerateDepsFile.ExecuteCore() [/go/src/agones.dev/agones/sdks/csharp/sdk/csharp-sdk.csproj]
/usr/share/dotnet/sdk/2.2.402/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.targets(129,5): error MSB4018:    at Microsoft.NET.Build.Tasks.TaskBase.Execute() [/go/src/agones.dev/agones/sdks/csharp/sdk/csharp-sdk.csproj]
/usr/share/dotnet/sdk/2.2.402/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.targets(129,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute() [/go/src/agones.dev/agones/sdks/csharp/sdk/csharp-sdk.csproj]
/usr/share/dotnet/sdk/2.2.402/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.targets(129,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskBuilder.ExecuteInstantiatedTask(ITaskExecutionHost taskExecutionHost, TaskLoggingContext taskLoggingContext, TaskHost taskHost, ItemBucket bucket, TaskExecutionMode howToExecuteTask) [/go/src/agones.dev/agones/sdks/csharp/sdk/csharp-sdk.csproj]
includes/sdk.mk:89: recipe for target 'run-sdk-command' failed
make[1]: *** [run-sdk-command] Error 1
includes/sdk.mk:85: recipe for target 'run-sdk-command-csharp' failed
```

File: examples/crd-client/main.go [hunk 6]
```diff
@@ -15,34 +15,51 @@
...
 	"k8s.io/client-go/rest"
 )
 
 const (
-	gameServerImage = "GAMESERVER_IMAGE"
+	gameServerImage      = "GAMESERVER_IMAGE"
+	isHelmTest           = "IS_HELM_TEST"
+	gameserversNamespace = "GAMESERVERS_NAMESPACE"
 
 	defaultImage = "gcr.io/agones-images/udp-server:0.19"
```
Comment (markmandel) [inline examples/crd-client/main.go:39]: What was the motivation for hardcoding the image here? You were getting it as an environment variable, and it looks like you are still passing it in from the yaml below. 

Seems like an env/arg value would be more flexible long term?

**Pull request merged**